lib/src/api_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,629 @@
+      if (element.enclosingElement3 is ExtensionElement) {
+        parentElement = element.enclosingElement3!;
+      } else if (node.realTarget.staticType is FunctionType) {
+        // do not allow the parent to be a Function
+        // this breaks the assertion node.realTarget.staticType?.element2 != null
+        return;
+      } else {
+        parentElement = node.realTarget.staticType?.element2! ??
+            (node.realTarget is ExtensionOverride
+                    ? (node.realTarget as ExtensionOverride).extensionName
+                    : node.realTarget as Identifier)
+                .staticElement!;
+      }
+      dependencyMetadata = identifyDependencyName(element: parentElement);
+    } on AnalysisException {
+      // do not continue if this invocation is unfit for analysis
+      return;
+    }
+
+    late bool constraintIssue;
+
+    // differentiate between class methods and top-level functions
+    if (parentElement is ClassElement) {
+      // ignore generics
+      if (parentElement.typeParameters.isNotEmpty) {
+        return;
+      }
+      constraintIssue = element is MethodElement
+          ? !dependencyMetadata.packageShape.containsMethodWithName(
+              parentElement.name, metadata.identifierName)
+          : !dependencyMetadata.packageShape.containsPropertyWithName(
+              parentElement.name, metadata.identifierName);
+    } else if (parentElement is ExtensionElement) {
+      // ignore generics
+      if (parentElement.typeParameters.isNotEmpty) {
+        return;
+      }
+      constraintIssue = element is MethodElement
+          ? !dependencyMetadata.packageShape.containsExtensionMethodWithName(
+              parentElement.name!, metadata.identifierName)
+          : !dependencyMetadata.packageShape.containsExtensionPropertyWithName(
+              parentElement.name!, metadata.identifierName);
+    } else {
+      // context.warning('Subclass ${parentElement.toString()} of parentElement (method/function) is not supported.');
+      return;
+    }
+
+    issues[metadata.elementId] = constraintIssue
+        ? PotentialLowerBoundConstraintIssue(
+            dependencyPackageName: dependencyMetadata.packageName,
+            constraint:
+                context.dependencies[dependencyMetadata.packageName]!.version,
+            currentVersion:
+                context.findInstalledVersion(dependencyMetadata.packageName),
+            lowestVersion:
+                Version.parse(dependencyMetadata.packageShape.version),
+            identifier: metadata.identifierName,
+            parentName: parentElement.name!,
+            kind: element.kind.toKind(),
+            parentKind: parentElement.kind.toParentKind(),
+            references: [metadata.span],
+          )
+        : null;
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    // an invocation of a top-level function or a class method
+    super.visitMethodInvocation(node);
```

Comment:
I'm doing it this way because I can `return` in case something cannot be resolved or the analysis reaches unexpected syntax - would it be better to do the super call last and use something like try-finally instead of returning?

Reply:
ah, good point about early returns.

If the order matters as I suggest, then you should probably rewrite to avoid early returns (a finally block would also be invoked in case of an exception, not really what you want).

Reply:
what makes you believe that the order might matter in some cases?

Reply:
It just feels more consistent to get the list of issues in reading order (from top->bottom, left->right.) I guess it doesn't really matter.